design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Check for existing 'next private key' Secret resources:
+            * If one matches, set the `status.nextPrivateKeySecretName` field.
+              This handles cache inconsistencies when we observe Secret
+              creation before updating `status.nextPrivateKeySecretName`.
+            * Otherwise, generate a new private key according to
+              `spec.privateKey` and store it in a new Secret resource.
+              Persist the name of the Secret as `status.nextPrivateKeySecretName`.
+    * If the `status.nextPrivateKeySecretName` field is set:
+        * If the Secret contains key data that matches `spec.privateKey`:
+            * do nothing
+        * If the Secret is not 'owned' by the Certificate, do nothing and log
+          an Event to inform the user.
+        * If the Secret is not labelled as a 'next private key', do nothing and
+          log an Event to inform the user.
+        * If the Secret does not exist:
+            * Generate a new private key according to `spec.privateKey` and
+              create a Secret resource with the given name.
+        * If the secret contains key data that does not match `spec.privateKey`:
+            * Generate a new private key according to `spec.privateKey` and
+              store it in the Secret resource.
+
+* If the `Issuing` condition is `False` or not set:
+    * Delete all owned Secret resources with the `cert-manager.io/next-private-key: "true`
+    * Ensure `status.nextPrivateKeySecretName` is unset - we may want to
+      consider not doing this in case a user has manually specified this field
+      and pointed it at an 'un-owned' Secret. This depends whether we want to
+      support this as a mode of operation.
+
+When creating a 'next private key' Secret resource, the
+`cert-manager.io/next-private-key: "true` annotation is added as well as an
+OwnerReference to the Certificate resource.
+
+Private keys generated by the key manager will be encoded in PKCS#8 format for
+ease of interoperability. The `issuing` controller will encode the resulting
+key-pair into the format requested by the user once the request has been
+completed.
+
+The introduction of this dedicated controller also means we can implement
+private key rotation when a certificate is re-issued/renewed.
+This is a welcome new feature, but in some cases a user may want to pin the
+private key used for a key-pair (as is the default and only supported behaviour
+prior to implementing this design).
+To continue to enable this, the `spec.privateKey.rotationPolicy` field
+controls how the next private key should be sourced. It supports two values:
+
+* `Always`: a new private key will be generated on every re-issuance.
+  This includes renewals as well as changes to the `spec.privateKey` and other
+  `spec` fields.
+* `Never`: private keys will never been regenerated and must be provided by the
+  user.
```

Comment:
I'm not sure whether `Never` rotating implies that the key *has* to be provided by the user, I would have expected a field on the `spec.privateKey` for setting a private key secret name, that then if non existent, would be created by cert-manager and then never updated if the policy was `Never`. This may be complicating things too much thought 😅 

Alternatively, there could be a third option `OnUpdate` that creates the secret and then only changes it if there is a spec change?

Reply:
If the policy were set to `Never` and an existing private key *does* exist, and doesn't match the requirements specified on the Certificate resource (i.e. wrong keySize or keyFormat), what would you expect the controller to do?

Reply:
What about in the context of using `spec.privateKey.pinnedSecretName`? Same behaviour? https://github.com/jetstack/cert-manager/pull/2753#discussion_r399343365